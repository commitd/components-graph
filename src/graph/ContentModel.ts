import { v4 } from 'uuid'
import {
  ModelAttributeSet,
  ModelEdge,
  ModelGraphData,
  ModelNode,
} from './types'

export class ContentModel {
  public static fromRaw(data: ModelGraphData): ContentModel {
    let model = Object.values(data.nodes).reduce(
      (acc, node) => acc.addNode(node),
      ContentModel.createEmpty()
    )
    model = Object.values(data.edges).reduce(
      (acc, edge) => acc.addEdge(edge),
      model
    )
    return model
  }

  public static createEmpty(): ContentModel {
    return new ContentModel({}, {})
  }

  constructor(
    readonly nodes: Record<string, ModelNode>,
    readonly edges: Record<string, ModelEdge>
  ) {
    this.nodes = nodes
    this.edges = edges
  }

  containsNode(id: string): boolean {
    return this.getNode(id) != null
  }

  containsEdge(id: string): boolean {
    return this.getEdge(id) != null
  }

  getNode(id: string): ModelNode | undefined {
    return this.nodes[`${id}`]
  }

  private getExistingNode(id: string): ModelNode {
    const node = this.getNode(id)
    if (node == null) {
      throw new Error(`Node [${id}] does not exist`)
    }
    return node
  }

  getEdgesLinkedToNode(nodeId: string): ModelEdge[] {
    return Object.values(this.edges).filter(
      (e) => e.target === nodeId || e.source === nodeId
    )
  }

  addNode(node: Partial<ModelNode>): ContentModel {
    if (node.id != null && this.containsNode(node.id)) {
      throw new Error(`Cannot add node already in graph (${node.id})`)
    }
    const newNode: ModelNode = {
      id: node.id ?? v4(),
      attributes: node.attributes ?? {},
      ...node,
    }
    const nodes = { ...this.nodes, [newNode.id]: newNode }
    return new ContentModel(nodes, this.edges)
  }

  editNode(node: ModelNode): ContentModel {
    this.getExistingNode(node.id)
    return new ContentModel(
      { ...this.nodes, ...{ [node.id]: node } },
      this.edges
    )
  }

  addNodeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    const node = this.getExistingNode(id)
    const newAttributes = {
      ...node.attributes,
      ...{ [attributeName]: attributeValue },
    }
    const changedNode = {
      ...this.getExistingNode(id),
      ...{ attributes: newAttributes },
    }
    return this.editNode(changedNode)
  }

  /**
   * Remove node and associated edges
   * @param id id of the node
   */
  removeNode(id: string): ContentModel {
    const node = this.getNode(id)
    if (node == null) {
      return this
    }
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    let content: ContentModel = this
    this.getEdgesLinkedToNode(id).forEach((e) => {
      content = content.removeEdge(e.id)
    })
    return content.removeOrphanNode(id)
  }

  private removeOrphanNode(id: string): ContentModel {
    const withoutNode = { ...this.nodes }
    delete withoutNode[`${id}`]
    return new ContentModel(withoutNode, this.edges)
  }

  editNodeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    if (this.getExistingNode(id).attributes[`${attributeName}`] == null) {
      throw new Error(`Node [${id}] does not have attribute ${attributeName}`)
    }
    return this.addNodeAttribute(id, attributeName, attributeValue)
  }

  removeNodeAttribute(id: string, attributeName: string): ContentModel {
    const node = this.getExistingNode(id)
    const remainingAttributes = { ...node.attributes }
    delete remainingAttributes[`${attributeName}`]
    const n = { ...node, ...{ attributes: remainingAttributes } }
    return this.editNode(n)
  }

  addEdge(
    edge: Omit<ModelEdge, 'id' | 'attributes'> & {
      id?: string
      attributes?: ModelAttributeSet
    }
  ): ContentModel {
    // check source and target exist
    this.getExistingNode(edge.source)
    this.getExistingNode(edge.target)
    if (edge.id != null && this.containsEdge(edge.id)) {
      throw new Error(`Cannot add edge already in graph (${edge.id})`)
    }
    const newEdge: ModelEdge = {
      id: edge.id ?? v4(),
      attributes: edge.attributes ?? {},
      ...edge,
    }
    const edges = { ...this.edges, [newEdge.id]: newEdge }
    return new ContentModel(this.nodes, edges)
  }

  getEdge(id: string): ModelEdge | undefined {
    return this.edges[`${id}`]
  }

  private getExistingEdge(id: string): ModelEdge {
    const edge = this.getEdge(id)
    if (edge == null) {
      throw new Error(`Edge [${id}] does not exist`)
    }
    return edge
  }

  editEdge(edge: ModelEdge): ContentModel {
    // ensure edge exists
    this.getExistingEdge(edge.id)
    return new ContentModel(this.nodes, {
      ...this.edges,
      ...{ [edge.id]: edge },
    })
  }

  addEdgeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    const edge = this.getExistingEdge(id)
    const newAttributes = {
      ...edge.attributes,
      ...{ [attributeName]: attributeValue },
    }
    const changedEdge = {
      ...this.getExistingEdge(id),
      ...{ attributes: newAttributes },
    }
    return this.editEdge(changedEdge)
  }

  editEdgeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    if (this.getExistingEdge(id).attributes[`${attributeName}`] == null) {
      throw new Error(`Edge [${id}] does not have attribute ${attributeName}`)
    }
    return this.addEdgeAttribute(id, attributeName, attributeValue)
  }

  removeEdgeAttribute(id: string, attributeName: string): ContentModel {
    const edge = this.getExistingEdge(id)
    const remainingAttributes = { ...edge.attributes }
    delete remainingAttributes[`${attributeName}`]
    const e = { ...edge, ...{ attributes: remainingAttributes } }
    return this.editEdge(e)
  }

  removeEdge(id: string): ContentModel {
    const edge = this.getEdge(id)
    if (edge == null) {
      return this
    }
    const withoutEdge = { ...this.edges }
    delete withoutEdge[`${id}`]
    return new ContentModel(this.nodes, withoutEdge)
  }
}
