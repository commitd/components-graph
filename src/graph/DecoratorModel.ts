import { colors, Theme } from '@committed/components'
import {
  DecoratedEdge,
  DecoratedNode,
  DecorationFunction,
  EdgeDecoration,
  EdgeDecorationFunction,
  EdgeDecorator,
  ItemDecoration,
  ModelEdge,
  ModelNode,
  NodeDecoration,
  NodeDecorationFunction,
  NodeDecorator,
} from './types'

export class DecoratorModel {
  static readonly DEFAULT_NODE_DECORATION: NodeDecoration = {
    shape: 'ellipse',
    color: '#FFc526',
    label: '',
    size: 25,
    opacity: 1,
    strokeColor: '#3E3E3E',
    strokeSize: 2,
  }

  static readonly DEFAULT_EDGE_DECORATION: EdgeDecoration = {
    color: '#FF0000',
    size: 2,
    label: '',
    sourceArrow: false,
    targetArrow: false,
    opacity: 1,
    style: 'solid',
  }

  static readonly idAsLabelNode: NodeDecorator = (item) => {
    return { label: item.id }
  }

  static readonly idAsLabelNodeEdge: EdgeDecorator = (item) => {
    return { label: item.id }
  }

  static createDefault(
    options: {
      nodeDecorators?: NodeDecorator[]
      nodeDefaults?: Partial<NodeDecoration> | NodeDecorationFunction
      edgeDecorators?: EdgeDecorator[]
      edgeDefaults?: Partial<EdgeDecoration> | EdgeDecorationFunction
    } = {}
  ): DecoratorModel {
    return new DecoratorModel(
      options.nodeDecorators ?? [],
      options.nodeDefaults ?? {},
      options.edgeDecorators ?? [],
      options.edgeDefaults ?? {}
    )
  }

  constructor(
    private readonly nodeDecorators: NodeDecorator[],
    private readonly nodeDefaults:
      | Partial<NodeDecoration>
      | NodeDecorationFunction,
    private readonly edgeDecorators: EdgeDecorator[],
    private readonly edgeDefaults:
      | Partial<EdgeDecoration>
      | EdgeDecorationFunction
  ) {
    this.nodeDecorators = nodeDecorators
    this.nodeDefaults = nodeDefaults
    this.edgeDecorators = edgeDecorators
    this.edgeDefaults = edgeDefaults
  }

  /**
   * Get the decoration that is applied to all nodes, before anything is overridden by decorators
   *
   * Use this to apply blanket styling to nodes to minimise the performance impact of per-node styling
   */
  getDefaultNodeDecorator(): DecorationFunction<
    NodeDecoration,
    NodeDecoration
  > {
    return (theme: Theme) => ({
      ...DecoratorModel.DEFAULT_NODE_DECORATION,
      ...{
        strokeColor: theme.palette.text.secondary,
        color: colors.committedYellow[500],
      },
      ...DecoratorModel.wrapDecorator(this.nodeDefaults)(theme),
    })
  }

  /**
   * Get the decoration that is applied to all edges, before anything is overridden by decorators
   *
   * Use this to apply blanket styling to edges to minimise the performance impact of per-edge styling
   */
  getDefaultEdgeDecorator(): DecorationFunction<
    EdgeDecoration,
    EdgeDecoration
  > {
    return (theme: Theme) => ({
      ...DecoratorModel.DEFAULT_EDGE_DECORATION,
      ...{
        color: theme.palette.text.secondary,
      },
      ...DecoratorModel.wrapDecorator(this.edgeDefaults)(theme),
    })
  }

  /**
   * Adds a function to the given nodes that provides the decoration overrides for the node
   * @param modelNodes to be decorated
   */
  getDecoratedNodes(modelNodes: ModelNode[]): DecoratedNode[] {
    return Object.values(modelNodes).map((node) => {
      return {
        getDecorationOverrides: (theme: Theme) => ({
          ...this.getNodeDecorationOverrides(node, theme),
        }),
        ...node,
      }
    })
  }

  /**
   * Adds a function to the given edges that provides the decoration overrides for the edge
   * @param modelEdges to be decorated
   */
  getDecoratedEdges(modelEdges: ModelEdge[]): DecoratedEdge[] {
    return Object.values(modelEdges).map((edge) => {
      return {
        getDecorationOverrides: (theme: Theme) => ({
          ...this.getEdgeDecorationOverrides(edge, theme),
        }),
        ...edge,
      }
    })
  }

  private getNodeDecorationOverrides(
    node: ModelNode,
    theme: Theme
  ): Partial<NodeDecoration> {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { id, attributes, ...nodeStyle } = node
    const decor = Object.assign(
      {},
      ...this.nodeDecorators.map((d) => d(node, theme))
    ) as Partial<NodeDecoration>
    return {
      ...decor,
      ...nodeStyle,
    }
  }

  private getEdgeDecorationOverrides(
    edge: ModelEdge,
    theme: Theme
  ): Partial<EdgeDecoration> {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { id, attributes, source, target, ...edgeStyle } = edge
    const decor = Object.assign(
      {},
      ...this.edgeDecorators.map((d) => d(edge, theme))
    ) as Partial<EdgeDecoration>
    return {
      ...decor,
      ...edgeStyle,
    }
  }

  private static isFunction<T extends ItemDecoration>(
    decorator: Partial<T> | DecorationFunction<T>
  ): decorator is DecorationFunction<T> {
    return typeof decorator === 'function'
  }

  private static wrapDecorator<T extends ItemDecoration>(
    decorator: Partial<T> | DecorationFunction<T>
  ): DecorationFunction<T> {
    if (DecoratorModel.isFunction(decorator)) {
      return decorator
    } else {
      return () => decorator
    }
  }
}
