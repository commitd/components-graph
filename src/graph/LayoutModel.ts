import { GraphLayout, GraphLayoutAlgorithm, PresetGraphLayout } from './types'

export class LayoutModel {
  private readonly layout: GraphLayout
  private readonly invalidated: boolean
  private readonly algorithm?: GraphLayoutAlgorithm

  static createDefault(): LayoutModel {
    return new LayoutModel('force-directed', true)
  }

  constructor(
    layout: GraphLayout,
    invalidated = true,
    algorithm?: GraphLayoutAlgorithm
  ) {
    this.layout = layout
    this.invalidated = invalidated
    this.algorithm = algorithm
  }

  getLayout(): GraphLayout {
    return this.layout
  }

  getLayoutAlgorithm(): GraphLayoutAlgorithm | undefined {
    return this.algorithm
  }

  isDirty(): boolean {
    return this.invalidated
  }

  validate(): LayoutModel {
    if (!this.invalidated) {
      return this
    } else {
      return new LayoutModel(this.layout, false, this.algorithm)
    }
  }

  invalidate(): LayoutModel {
    if (this.invalidated) {
      return this
    } else {
      return new LayoutModel(this.layout, true, this.algorithm)
    }
  }

  presetLayout(layout: PresetGraphLayout): LayoutModel {
    return new LayoutModel(layout, true, undefined)
  }

  customLayout(algorithm: GraphLayoutAlgorithm): LayoutModel {
    return new LayoutModel('custom', true, algorithm)
  }
}
