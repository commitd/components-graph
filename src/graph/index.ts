export * from './ContentModel'
export * from './GraphModel'
export * from './DecoratorModel'
export * from './LayoutModel'
export * from './SelectionModel'
export * from './data'
export * from './types'
export * from './renderer'
