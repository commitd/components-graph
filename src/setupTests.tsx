import { ThemeProvider } from '@committed/components'
import { StylesProvider } from '@material-ui/styles'
import '@testing-library/jest-dom/extend-expect'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GenerateId } from 'jss'
import React from 'react'

const generateClassName: GenerateId = (rule, styleSheet) => {
  const prefix =
    styleSheet === undefined || styleSheet.options.classNamePrefix === undefined
      ? ''
      : styleSheet.options.classNamePrefix
  return `${prefix}-${rule.key}`
}

const LightTheme: React.FC = ({ children }) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="light">{children}</ThemeProvider>
  </StylesProvider>
)

const DarkTheme: React.FC = ({ children }) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="dark">{children}</ThemeProvider>
  </StylesProvider>
)

export const renderPlain = render

export const renderLight = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: LightTheme, ...options })

export const renderDark = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: DarkTheme, ...options })

// re-export everything
export * from '@testing-library/react'
export { userEvent }
