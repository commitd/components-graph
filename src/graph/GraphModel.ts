import { ContentModel } from './ContentModel'
import { DecoratorModel } from './DecoratorModel'
import { LayoutModel } from './LayoutModel'
import { SelectionModel } from './SelectionModel'
import {
  DecoratedEdge,
  DecoratedNode,
  GraphCommand,
  ModelEdge,
  ModelNode,
} from './types'

/** The GraphModel is a declarative, immutable definition of graph state
 
The GraphModel defines:
- Nodes/edges (ContentModel)
	- No requirement for ontology, typing etc
	- Attributes
- Decoration (DecoratorModel)
	- May define custom node/edge Decorators based on any property on the node/edge
- Selection (SelectionModel)
	- Multi-select
- Graph layout (LayoutModel)
	- Force Layout / Cola JS
	- Circle
	- Grid
	- Custom (supplying a function)
- CommandQueue
	- Trigger behaviour from outside the GraphModel/React component
	- Current commands: zoom in/out, relayout, refit
 */
export class GraphModel {
  private readonly contentModel: ContentModel
  private readonly selectionModel: SelectionModel
  private readonly decoratorModel: DecoratorModel
  private readonly layoutModel: LayoutModel
  private readonly commandQueue: GraphCommand[]

  constructor(
    contentModel: ContentModel,
    options: {
      selectionModel?: SelectionModel
      decoratorModel?: DecoratorModel
      layoutModel?: LayoutModel
      commandQueue?: GraphCommand[]
    } = {}
  ) {
    this.contentModel = contentModel
    this.decoratorModel =
      options.decoratorModel ?? DecoratorModel.createDefault()
    this.layoutModel = options.layoutModel ?? LayoutModel.createDefault()
    this.selectionModel =
      options.selectionModel ?? SelectionModel.createDefault()
    this.commandQueue = options.commandQueue ?? []
  }

  static applyContent(
    graphModel: GraphModel,
    contentModel: ContentModel
  ): GraphModel {
    return new GraphModel(contentModel, {
      ...graphModel.getOptions(),
      selectionModel: GraphModel.reconcileSelection(
        contentModel,
        graphModel.getSelection()
      ),
    })
  }

  static applySelection(
    graphModel: GraphModel,
    selectionModel: SelectionModel
  ): GraphModel {
    return new GraphModel(graphModel.getCurrentContent(), {
      ...graphModel.getOptions(),
      selectionModel: GraphModel.reconcileSelection(
        graphModel.getCurrentContent(),
        selectionModel
      ),
    })
  }

  static applyLayout(
    graphModel: GraphModel,
    layoutModel: LayoutModel
  ): GraphModel {
    return new GraphModel(graphModel.contentModel, {
      ...graphModel.getOptions(),
      layoutModel,
      selectionModel: graphModel.getSelection(),
    })
  }

  static createEmpty(): GraphModel {
    return new GraphModel(ContentModel.createEmpty())
  }

  static createWithContent(contentModel: ContentModel): GraphModel {
    return new GraphModel(contentModel)
  }

  getCurrentContent(): ContentModel {
    return this.contentModel
  }

  getDecorators(): DecoratorModel {
    return this.decoratorModel
  }

  // get the current fully decorated nodes
  get nodes(): DecoratedNode[] {
    return this.decoratorModel.getDecoratedNodes(
      Object.values(this.contentModel.nodes)
    )
  }

  // get the current fully decorated edges
  get edges(): DecoratedEdge[] {
    return this.decoratorModel.getDecoratedEdges(
      Object.values(this.contentModel.edges)
    )
  }

  get selectedNodes(): DecoratedNode[] {
    return this.decoratorModel.getDecoratedNodes(
      Array.from(this.getSelection().nodes)
        .map((n) => this.contentModel.getNode(n))
        .filter((n) => n != null)
        .map((n) => n as ModelNode)
    )
  }

  get selectedEdges(): DecoratedEdge[] {
    return this.decoratorModel.getDecoratedEdges(
      Array.from(this.getSelection().edges)
        .map((e) => this.contentModel.getEdge(e))
        .filter((e) => e != null)
        .map((e) => e as ModelEdge)
    )
  }

  getNode(id: string): DecoratedNode | undefined {
    const node = this.contentModel.getNode(id)
    if (node == null) {
      return node
    }
    return this.decoratorModel.getDecoratedNodes([node])[0]
  }

  getEdge(id: string): DecoratedEdge | undefined {
    const edge = this.contentModel.getEdge(id)
    if (edge == null) {
      return edge
    }
    return this.decoratorModel.getDecoratedEdges([edge])[0]
  }

  getCurrentLayout(): LayoutModel {
    return this.layoutModel
  }

  getSelection(): SelectionModel {
    return this.selectionModel
  }

  getCommands(): GraphCommand[] {
    return this.commandQueue
  }

  pushCommand(command: GraphCommand): GraphModel {
    return new GraphModel(this.contentModel, {
      ...this.getOptions(),
      commandQueue: [...this.commandQueue, command],
    })
  }

  clearCommands(): GraphModel {
    if (this.commandQueue.length === 0) {
      return this
    }
    return new GraphModel(this.contentModel, {
      ...this.getOptions(),
      commandQueue: [],
    })
  }

  private getOptions(): {
    selectionModel: SelectionModel
    decoratorModel: DecoratorModel
    layoutModel: LayoutModel
    commandQueue: GraphCommand[]
  } {
    return {
      selectionModel: this.selectionModel,
      decoratorModel: this.decoratorModel,
      layoutModel: this.layoutModel,
      commandQueue: this.commandQueue,
    }
  }

  private static reconcileSelection(
    contentModel: ContentModel,
    selectionModel: SelectionModel
  ): SelectionModel {
    const nodes = Array.from(selectionModel.nodes).filter((n) =>
      contentModel.containsNode(n)
    )
    const edges = Array.from(selectionModel.edges).filter((n) =>
      contentModel.containsEdge(n)
    )
    return new SelectionModel(nodes, edges)
  }
}
