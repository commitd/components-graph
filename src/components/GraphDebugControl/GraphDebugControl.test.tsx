import React from 'react'
import { GraphDebugControl } from '.'
import { GraphModel } from '../../graph/GraphModel'
import { renderDark, renderLight, userEvent } from '../../setupTests'

it('renders light', () => {
  const { asFragment } = renderLight(
    <GraphDebugControl
      model={GraphModel.createEmpty()}
      onChange={jest.fn()}
      onReset={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders dark', () => {
  const { asFragment } = renderDark(
    <GraphDebugControl
      model={GraphModel.createEmpty()}
      onChange={jest.fn()}
      onReset={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

test.each([
  ['Add Node'],
  ['Add 10 Nodes'],
  ['Add 100 Nodes'],
  ['Add Edge'],
  ['Add 10 Edges'],
  ['Add 100 Edges'],
  ['Remove Random Node'],
  ['Remove Random Edge'],
  ['Layout'],
])('button label %s changes graph', (label) => {
  const onChange = jest.fn()
  const { getByRole } = renderLight(
    <GraphDebugControl
      model={GraphModel.createEmpty()}
      onChange={onChange}
      onReset={jest.fn()}
    />
  )

  userEvent.click(getByRole('button', { name: label }))
  expect(onChange).toHaveBeenCalled()
})

it('reset', () => {
  const onReset = jest.fn()
  const { getByRole } = renderLight(
    <GraphDebugControl
      model={GraphModel.createEmpty()}
      onChange={jest.fn()}
      onReset={onReset}
    />
  )

  userEvent.click(getByRole('button', { name: 'Reset' }))
  expect(onReset).toHaveBeenCalled()
})

it('layout', () => {
  const onChange = jest.fn()
  const { getByRole } = renderLight(
    <GraphDebugControl
      model={GraphModel.createEmpty()}
      onReset={jest.fn()}
      onChange={onChange}
    />
  )

  userEvent.click(getByRole('button', { name: 'force-directed' }))
  userEvent.click(getByRole('option', { name: 'circle' }))
  expect(onChange).toHaveBeenCalled()
})
