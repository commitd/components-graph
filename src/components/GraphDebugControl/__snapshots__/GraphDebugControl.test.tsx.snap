// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dark 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add Node
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 10 Nodes
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 100 Nodes
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add Edge
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 10 Edges
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 100 Edges
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Remove Random Node
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Remove Random Edge
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Layout
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Reset
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root ForwardRef(withBox)-root ForwardRef(withBox)-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink="true"
        >
          Layout
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root makeStyles-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <div
            aria-haspopup="listbox"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
            role="button"
            tabindex="0"
          >
            force-directed
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            tabindex="-1"
            value="force-directed"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders light 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add Node
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 10 Nodes
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 100 Nodes
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add Edge
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 10 Edges
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add 100 Edges
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Remove Random Node
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Remove Random Edge
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Layout
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Reset
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root ForwardRef(withBox)-root ForwardRef(withBox)-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink="true"
        >
          Layout
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root makeStyles-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <div
            aria-haspopup="listbox"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
            role="button"
            tabindex="0"
          >
            force-directed
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            tabindex="-1"
            value="force-directed"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
